### ISSUE TITLE
Wallet: Whitelist and Spending Limit Bypass

### ISSUE THREAD

**Issue by mischat** [tokencard/contracts#491]
In the event of an end-user key disclosure, an attacker can bypass the wallet’s address whitelist through the `executeTransaciton{}` function.

The `executeTransaction()` function blacklists the ERC20 `transfer()` and `approve()` methods insofar as it continues to enforce spending limits and whitelist rules on for those methods. However, those are not the only methods that ERC20 tokens can implement for transferring tokens. For example, many tokens implement methods like `increaseApproval()` and `decreaseApproval()`, which would still allow an attacker to create an allowance for an attacker-controlled account without respect to whitelist and spending limit rules.

Since many ERC20 tokens have mutable APIs (i.e. they use upgradeable contracts), it it may not be possible to guarantee that the `executeTransaction()` functionality is ever completely safe, even if it were to incorporate a much larger blacklist of functions.

**Comment by mischat** [comment]
This is addressed in PR https://github.com/tokencard/contracts/pull/486

**Comment by mischat** [comment]
This was an excellent find, first found by sunfish, we are really happy with this, thank you.

**Comment by mischat** [comment]
sunfish aka Phil :)

**Comment by mischat** [comment]
We have moved away from the concept of "blacklisting functions" in `executeTransaction()` as identified by the audit. 

We are in a world now, whereby we whitelist `transfer`, `transferFrom`, `burn`, and `approve` based on ERC20 standard. We protect tokens in our `tokenWhitelist` by enforcing the daily spend limit by allowing only the aforementioned ERC20 functions. We also protect ETH transfers of our users' wallets.  All smart contract operations for contracts not within our `tokenWhitelist` are permissable.

**Comment by mischat** [comment]
The final point on this by phil is as follows : 

```
The strategy of whitelisting the token addresses *and*
methods seems fine, provided that you've manually verified
that the tokens in question have more-or-less "normal"
behavior in their ERC20 transfer/approve/etc. methods.
For upgrade-able tokens, you'll have to do a bit of
due diligence on an ongoing basis, but that was already
the case anyway.

It's probably worth noting in your user documentation
that the wallet doesn't have safety features that guard
_unsupported_ tokens, etc. (For example, if someone were
to own an ENS address through their wallet, that would still
be affected by a key disclosure. So users should probably
be advised against doing something like that.)

It's also worth noting that method whitelisting only works
in concert with the token address whitelist. There aren't
enough bits in a method signature (32) to ensure that someone
couldn't produce a contract with another method with the same
signature. (You can find a 32-bit sha3 prefix preimage in a few
minutes on conventional hardware.) So the presence of your
token address whitelist is absolutely instrumental in ensuring
that things work correctly, since it implies that you "trust"
the ABI of the contracts being called.
```

**Comment by mischat** [comment]
This is fixed and merged in version version 2.2.0
